x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java

```diff
@@ -46,16 +46,36 @@
 import org.elasticsearch.xpack.core.ccr.action.ResumeFollowAction;
 
 import java.util.ArrayList;
+import java.util.Comparator;
 import java.util.List;
 import java.util.Locale;
 import java.util.Objects;
 import java.util.function.BiConsumer;
 import java.util.stream.Collectors;
 
+import static org.elasticsearch.cluster.metadata.DataStream.BACKING_INDEX_PREFIX;
+
 public final class TransportPutFollowAction extends TransportMasterNodeAction<PutFollowAction.Request, PutFollowAction.Response> {
 
     private static final Logger logger = LogManager.getLogger(TransportPutFollowAction.class);
 
+    /**
+     * Comparator used to sort the backing indices of the local data stream when following a backing index that's not currently part of
+     * the local data stream.
+     * The backing indices usually contain `.ds-` (they usually start with this prefix, but not always as shrunk indices will have the
+     * `shrink-` prefix for e.g.) but given the `modify data stream API` any index name can be a backing index.
+     * For this reason this comparator groups the indices that don't contain `.ds-` in the lower generations, and the ones that do
+     * contain `.ds-` are ordered by the original (provided) backing index name - the name starting with `.ds-`.
+     * The goal is to make sure the prefixed (usually read-only - shrink-, restored-, partial-) backing indices do not end up being the
+     * write index of the local data stream.
+     */
+    private static final Comparator<Index> LOCAL_DATA_STREAM_INDICES_COMPARATOR = Comparator.comparing(
+        (Index o) -> o.getName().contains(BACKING_INDEX_PREFIX) ? 1 : -1
```

Comment:
One last refinement I thought of is the unlikely edge case that there is a `.ds-` prefixed backing index in the data stream that might have been moved over from a different data stream. Maybe we should parameterize this comparator instance with the data stream name and use it in the partition step: 

```
private String partitionByIndexBaseName = BACKING_INDEX_PREFIX + dataStreamName;
...
o.getName().contains(partitionByIndexBaseName)
```
instead of just using the `BACKING_INDEX_PREFIX` to partition indices? It's a very unlikely edge case but I think it seals things up pretty well while we think about a better long term fix than sorting by index name.